test/compiler/inline.jl

```diff
@@ -1279,3 +1279,89 @@ end
 # Test that inlining doesn't accidentally delete a bad return_type call
 f_bad_return_type() = Core.Compiler.return_type(+, 1, 2)
 @test_throws MethodError f_bad_return_type()
+
+# Test that we can inline a finalizer for a struct that does not otherwise escape
+global total_deallocations::Int = 0
+
+mutable struct DoAllocNoEscape
+    function DoAllocNoEscape()
+        finalizer(new()) do this
+            global total_deallocations += 1
```

Comment:
This is technically an undefined data race and therefore invalid code. We don't yet support writing this correctly, which would hopefully soon be:
```julia
@atomic global total_deallocations += 1
```
such that it is able to be turned into a call like so:
```
modifyglobal!(@__MODULE__, :total_deallocations, +, 1)
```
so that this finalizer isn't relying on undefined behavior anymore.

In the meantime, perhaps this should use an AtomicRef instead of a global?

Reply:
Will switch this to an appropriately annotated ccall. Effect modeling for modifyfield! is a whole separate can of worms.